prompt_toolkit/input/win32.py

```diff
@@ -348,7 +348,23 @@ def _event_to_key_presses(self, ev):
             else:
                 result = KeyPress(u_char, u_char)
 
-        # Correctly handle Control-Arrow keys.
+        # First we handle Shift-Control-Arrow/Home/End (need to do this first)
+        if (ev.ControlKeyState & self.LEFT_CTRL_PRESSED or
+                ev.ControlKeyState & self.RIGHT_CTRL_PRESSED) and \
+                ev.ControlKeyState & self.SHIFT_PRESSED and result:
+            if result.key == Keys.Left:
+                result.key = Keys.ShiftControlLeft
+
+            if result.key == Keys.Right:
+                result.key = Keys.ShiftControlRight
+
+            if result.key == Keys.Home:
+                result.key = Keys.ShiftControlHome
+
+            if result.key == Keys.End:
+                result.key = Keys.ShiftControlEnd
+
+        # Correctly handle Control-Arrow/Home/End and Control-Insert keys.
         if (ev.ControlKeyState & self.LEFT_CTRL_PRESSED or
```

Comment:
I wonder whether this doesn't have to be an ``elif``? If both shift and control are pressed don't we end up in both this branch and the ``if` above. I think we replace ``Keys.ShiftControlLeft`` by ``ControlLeft`` like this. (Maybe I'm missing something.)

Reply:
Never mind. The assignment to ``result.key`` ensures that none of the other conditions satisfy. This is fine.